### DISCUSSION TITLE
Unexpected Verification Error in when calling smart_table::add

### DISCUSSION THREAD

**Discussion by 0x-j** [aptos-labs/aptos-developer-discussions#360] [Questions] [answered]
### Discord user ID

nguyendangduong3711 

### Describe your question in detail.

Asked from discord

Run this unit test

```rust
module dummy::test {
    use std::debug::print;
    use aptos_std::smart_table::{Self, SmartTable};
    #[test]
    fun test() {
        let table = smart_table::new<u64, u64>();
        smart_table::add<u64,u64>(&mut table, 1u64, 100u64);
      
        smart_table::clear(&mut table);
        let length = smart_table::length(&table);
        assert!(length == 0, 0);
        
        smart_table::destroy_empty(table);
    }
}
```

### What error, if any, are you getting?

```
native public fun serialized_size<MoveValue>(v: &MoveValue): u64;
INTERNAL TEST ERROR: Unexpected Verification Error
est was not expected to error, but it gave a MISSING_DEPENDENCY (code 1021) error with error message: "Missing Native Function `serialized_size`". Error originating in the module 0000000000000000000000000000000000000000000000000000000000000001::bcs rooted here
In this function in 0x1::bcs
stack trace:
   smart_table::add(/home/duongnd/.move/https___github_com_aptos-labs_aptos-core_git_main/aptos-move/framework/aptos-framework/../aptos-stdlib/sources/data_structures/smart_table.move:145)
   test::test(./sources/test.move:12)
```

### Which operating system are you using?

macOS

### Which SDK or tool are you using? (if any)

TypeScript SDK

**Comment by 0x-j** [comment]
I can reproduce this error when i'm using main branch of aptos-core for AptosFramework.
Commit `9c21cf58c931faafd6b4dff30583f34fb3643de8`

If I switched to mainnet branch of aptos-core, it works fine.
Commit `9802bd084aadcd3565e95ca7f72ff940212875e5`

**Reply by georgemitenkov** [reply]
Hey! We are aware of the issue and are fixing it. Which version of the CLI are you using? The problem is because the changes recently made to `main` are unstable, and the old CLIs might now be compatible with them.

**Reply by 0x-j** [reply]
thanks! im using cli 3.5.1

**Reply by IsaacCisnerosFT** [reply]
@georgemitenkov is there any open GitHub issue related to this? I am also running CLI v3.5.1 and experiencing similar behavior using devnet and testnet revisions. Running my tests with mainnet revision solves the problem for now but I would like to track the current resolution progress.

**Reply by MoonShiesty** [reply]
ditto on CLI v3.5.1

**Reply by georgemitenkov** [reply]
The issue is filed & tracked here: https://github.com/aptos-labs/aptos-core/issues/14240

The problem with the existing CLI is that the newest aptos framework release now contains a new native function to enhance calculation of serialised size of a Move value (cheaper gas, faster implementation) which broke compatibility with previously released CLIs. The not so user-friendly error message `MISSING_DEPENDENCY` is meant to say that the new native function cannot be found (which is the case for old CLIs). 

We are releasing the new CLI in around today/tomorrow which solves the issue. In the meantime, building CLI from `main` is a workaround:
```
cd aptos-core
cargo build -p aptos --release
```
Having encountered this issue, we are also looking into making error messages better and improving devex when working with the CLI.

**Comment by georgemitenkov** [comment, marked as answer]
Hey @MoonShiesty @IsaacCisnerosFT @0xaptosj! The new CLI (4.0.0) has been released which fixes the issue described here. If you update, things should work as normal.

https://github.com/aptos-labs/aptos-core/releases/tag/aptos-cli-v4.0.0
